src/librbd/Operations.cc

```diff
@@ -1119,34 +1126,50 @@ void Operations<I>::execute_snap_unprotect(const std::string &snap_name,
 }
 
 template <typename I>
-int Operations<I>::prepare_image_update() {
+int Operations<I>::prepare_image_update(bool request_lock) {
   assert(m_image_ctx.owner_lock.is_locked() &&
          !m_image_ctx.owner_lock.is_wlocked());
-  if (m_image_ctx.image_watcher == NULL) {
+  if (m_image_ctx.image_watcher == nullptr) {
     return -EROFS;
   }
 
   // need to upgrade to a write lock
   int r = 0;
-  bool trying_lock = false;
-  C_SaferCond ctx;
+
+  Context *ctx = util::create_async_context_callback(
+    m_image_ctx, util::create_context_callback<
+      C_InvokeAsyncRequest<I>,
+      &C_InvokeAsyncRequest<I>::handle_acquire_exclusive_lock>(this));
```

Comment:
@trociny How do I initialize `ctx` correctly here so I can send it to request_lock and try_lock, below?

Reply:
@smithfarm you don't need this. `C_SaferCond ctx` was perfectly good for this. What you only need is to replace:
```
+      if (request_lock) {
+        m_image_ctx.exclusive_lock->acquire_lock(&ctx);
+      } else {
+        m_image_ctx.exclusive_lock->try_acquire_lock(&ctx);
+      }
```
from the original patch with:
```
+      if (request_lock) {
+        m_image_ctx.exclusive_lock->request_lock(&ctx);
+      } else {
+        m_image_ctx.exclusive_lock->try_lock(&ctx);
+      }
```